### PR TITLE
Refactor Node class

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

- Messy error handling in Node
- This continues the fix in  #1243 and #1245.

### How was it fixed?

- Leave only the happy path in the function body, define different new exceptions, and aggregate all error handling in a function.
- Moving many util functions from Node to utils.py

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: trinity/protocol/bcc_libp2p/utils.py [hunk 1]
```diff
@@ -164,38 +487,71 @@ def make_rpc_v1_ssz_protocol_id(message_name: str) -> str:
...
+        await stream.close()
+        raise WriteMessageFailure() from error
+    except StreamReset as error:
+        raise WriteMessageFailure() from error
+
+
+async def _read_stream(stream: INetStream, len_payload: int, timeout: float) -> bytes:
+    try:
+        return await asyncio.wait_for(stream.read(len_payload), timeout)
+    except asyncio.TimeoutError as error:
```
Comment (NIC619) [inline trinity/protocol/bcc_libp2p/utils.py]: nit:
```suggestion
    except asyncio.TimeoutError:
```

File: trinity/protocol/bcc_libp2p/utils.py [hunk 2]
```diff
@@ -164,38 +487,71 @@ def make_rpc_v1_ssz_protocol_id(message_name: str) -> str:
...
+        await stream.close()
+        raise ReadMessageFailure() from error
+    except StreamReset as error:
+        raise ReadMessageFailure() from error
+
+
+async def _decode_uvarint_from_stream(stream: INetStream, timeout: float) -> None:
+    try:
+        return await asyncio.wait_for(decode_uvarint_from_stream(stream), timeout)
+    except asyncio.TimeoutError as error:
```
Comment (NIC619) [inline trinity/protocol/bcc_libp2p/utils.py]: nit:
```suggestion
    except asyncio.TimeoutError:
```

File: trinity/protocol/bcc_libp2p/exceptions.py [hunk 3]
```diff
@@ -1,22 +1,54 @@
...
 class RequestFailure(BaseLibp2pError):
-    pass
+    ...
+
+
+class PeerRespondedAnError(BaseLibp2pError):
+    ...
+
+
+class InvalidRequestSaidPeer(PeerRespondedAnError):
```
Comment (NIC619) [inline trinity/protocol/bcc_libp2p/exceptions.py]: I like the detailed naming and classification of the exceptions.

File: trinity/protocol/bcc_libp2p/node.py [hunk 4]
```diff
@@ -598,583 +531,121 @@ def _compare_chain_tip_and_finalized_epoch(self,
...
+            peer_id = interaction.peer_id
             try:
-                block = self.chain.get_canonical_block_by_slot(slot)
-            except BlockNotFound:
+                await interaction.read_request(Goodbye)
+            except ReadMessageFailure:
                 pass
-            else:
-                yield block
+            await self.disconnect_peer(peer_id)
```
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply trinity/protocol/bcc_libp2p/node.py]: I hacked the serializable object, so now the interaction will log the human-readable goodbye reason now. It looks like:
```
Received request Goodbye from 16Uiu2HAmNXgEPcuEr3U5p6vCyhaMkdGExmmhxvJct1nW2E3BsXbH  {'reason': <GoodbyeReasonCode.FAULT_OR_ERROR: 3>}
```

**Pull request merged**